File: ios/Classes/strategies/EventPermissionStrategy.m
//
// Created by Razvan Lung(long1eu) on 2019-02-15.
// Copyright (c) 2019 The Chromium Authors. All rights reserved.
//

#import "EventPermissionStrategy.h"


@implementation EventPermissionStrategy

- (PermissionStatus)checkPermissionStatus:(PermissionGroup)permission {
    if (permission == PermissionGroupCalendar) {
        return [EventPermissionStrategy permissionStatus:EKEntityTypeEvent];
    } else if (permission == PermissionGroupReminders) {
        return [EventPermissionStrategy permissionStatus:EKEntityTypeReminder];
    }
    
    return PermissionStatusUnknown;
}

- (ServiceStatus)checkServiceStatus:(PermissionGroup)permission {
    return ServiceStatusNotApplicable;
}

- (void)requestPermission:(PermissionGroup)permission completionHandler:(PermissionStatusHandler)completionHandler {
    PermissionStatus permissionStatus = [self checkPermissionStatus:permission];
    
    if (permissionStatus != PermissionStatusUnknown) {
        completionHandler(permissionStatus);
        return;
    }
    
    EKEntityType entityType;
    
    if (permission == PermissionGroupCalendar) {
        entityType = EKEntityTypeEvent;
    } else if (permission == PermissionGroupReminders) {
        entityType = EKEntityTypeReminder;
    } else {
        completionHandler(PermissionStatusUnknown);
        return;
    }
    
    EKEventStore *eventStore = [[EKEventStore alloc] init];
    [eventStore requestAccessToEntityType:entityType completion:^(BOOL granted, NSError *error) {
        if (granted) {
            completionHandler(PermissionStatusGranted);
        } else {
            completionHandler(PermissionStatusDenied);
        }
    }];
}

+ (PermissionStatus)permissionStatus:(EKEntityType)entityType {
    EKAuthorizationStatus status = [EKEventStore authorizationStatusForEntityType:entityType];
    
    switch (status) {
        case EKAuthorizationStatusNotDetermined:
            return PermissionStatusUnknown;
        case EKAuthorizationStatusRestricted:
            return PermissionStatusRestricted;
        case EKAuthorizationStatusDenied:
            return PermissionStatusDenied;
        case EKAuthorizationStatusAuthorized:
            return PermissionStatusGranted;
    }
    
    return PermissionStatusUnknown;
}

@end

File: ios/Classes/strategies/EventPermissionStrategy.swift
//
//  EventPermissions.swift
//  permission_handler
//
//  Created by Maurits van Beusekom on 26/07/2018.
//

import EventKit
import Foundation

class EventPermissionStrategy : NSObject, PermissionStrategy {

    func checkPermissionStatus(permission: PermissionGroup) -> PermissionStatus {
        if permission == PermissionGroup.calendar {
            return EventPermissionStrategy.getPermissionStatus(entityType: EKEntityType.event)
        } else if permission == PermissionGroup.reminders {
            return EventPermissionStrategy.getPermissionStatus(entityType: EKEntityType.reminder)
        }
        
        return PermissionStatus.unknown
    }
    
    private static func getPermissionStatus(entityType: EKEntityType) -> PermissionStatus {
        let status: EKAuthorizationStatus = EKEventStore.authorizationStatus(for: entityType)
        
        switch status {
        case EKAuthorizationStatus.authorized:
            return PermissionStatus.granted
        case EKAuthorizationStatus.denied:
            return PermissionStatus.denied
        case EKAuthorizationStatus.restricted:
            return PermissionStatus.restricted
        default:
            return PermissionStatus.unknown
            
        }
    }
    
    func checkServiceStatus(permission: PermissionGroup) -> ServiceStatus {
        return ServiceStatus.notApplicable
    }
    
    func requestPermission(permission: PermissionGroup, completionHandler: @escaping PermissionStatusHandler) {
        let permissionStatus = checkPermissionStatus(permission: permission)
        
        if permissionStatus != PermissionStatus.unknown {
            completionHandler(permissionStatus)
            return
        }
        
        var entityType: EKEntityType
        
        if permission == PermissionGroup.calendar {
            entityType = EKEntityType.event
        } else if permission == PermissionGroup.reminders {
            entityType = EKEntityType.reminder
        } else {
            completionHandler(PermissionStatus.unknown)
            return
        }
        
        let eventStore: EKEventStore = EKEventStore.init()
        eventStore.requestAccess(to: entityType) { (granted: Bool, error: Error?) in
            if granted {
                completionHandler(PermissionStatus.granted)
            } else {
                completionHandler(PermissionStatus.denied)
            }
        }
    }
}
